app/common/helpers/collections.py

```diff
@@ -73,11 +72,14 @@ def reference(self) -> str:
 
     @property
     def expression_context(self) -> ExpressionContext:
-        return ExpressionContext(
-            from_submission=immutabledict(
-                {mangle_question_id_for_context(uuid.UUID(k)): v for k, v in self.submission.data.items()}
-            )
-        )
+        submission_data = {
+            question.safe_qid: answer.model_dump()
+            for section in self.submission.collection.sections
+            for form in section.forms
+            for question in form.questions
+            if (answer := self.get_answer_for_question(question.id)) is not None
```

Comment:
This little refactor on its own might be worth calling out as a separate change! Very nice, I think this sets us up for consistent context data and we'll just need to work out how its sensible for dynamic forms to be doing this processing too